### DISCUSSION TITLE
Add new flag to reinstall dependencies (--clean or --phoenix)

### DISCUSSION THREAD

**Discussion by midudev** [npm/feedback#342] [CLI] [answered]
Add a new flag (`--clean` or `--phoenix`).

When added to `npm install`, it cleans the `node_modules` folders.

Normally we would do on package.json something like this:

```json
"phoenix": "rm -rf ./node_modules package-lock.json && npm install"
```

But with the new `workspaces` feature this is more complicated as a `node_modules` folder could be present on each `workspace` and you should be sure to clean them as well to avoid some troubles.

I think this flag could be useful as I see this is FAQ on some tweets and my community.

PS: Maybe you've already considered this but I've been not able to find something like this. If it's the case, sorry!

**Comment by ljharb** [comment]
When you have a lockfile present, `npm ci` already achieves this for node_modules. Does that cover your use case?

**Reply by targos** [reply]
Given the example script, I think their use case is a different one that I have too: install the latest compatible version of all dependencies (recursively) and generate a fresh lock file.
In my experience, removing node_modules and the lock file is the only way to do it cleanly.

**Reply by ljharb** [reply]
I agree with that.

@midudev what is your actual use case for needing this?

**Reply by midudev** [reply]
Thanks for the reply @ljharb! 👏

I didn't know that `npm ci` do that but the command is intended for being used on CI environments (also in my company we're not using lockfiles for several reasons but that's another topic). I see this flag more for local npm install usage as CI environment is usually a "clean" slate.

Anyway, let me share some use cases that I see on my company.

a) You haven't touched the repo for a time and you want to be sure you're installing latest versions for all packages. We're using in my company major versions (we use Compatible Versioning) and we usually remove node_modules to be sure we're installing latest versions. Specially for subdependencies.

b) When migrating from v6 to v7 keeping node_modules on subpackages could be problematic as packages are not being removed (as far as I tried last time).

c) Developer or tool has modified something on node_modules manually corrupting the dependencies (for example, some lazy dev didn't want to use link and instead modified files). Not cool, I know, but could happen. In workspaces fixing this is, for sure, needing to remove `node_modules` as could be tricky to find those files.

d) You want to be sure you're removing cache for packages like babel (thousand of files) but at the same time you want to be sure you're using latest compatible versions.

Now, I've done a quick search on Twitter about this and I found a tweet about some days ago regarding the same topic:
https://twitter.com/d4nidev/status/1382639135760977921

I know, we don't do Twitter Driven Development (😆) I only wanted to share that this is something that seems not to be something singular.

In our case, we're doing a workaround by creating a tool that deals with workspaces and we execute commands (commands not scripts) for each one. Still far from elegant. Also, `rm -rf` is... scary and not compatible with Windows. :D

That's why I'm wondering if a flag for `npm install` could make sense. :) So, no more `rm -rf`, more elegant and available for everyone.

Thanks!

**Reply by ljharb** [reply]
I'm very surprised your company isn't using lockfiles; it's a best practice for apps to have lockfiles in virtually every ecosystem i'm aware of that has the feature. In general, I'm not sure it would make any sense to design a feature in npm around an app that doesn't use a lockfile. That said, I don't personally believe packages should have lockfiles, and when I need a clean install in my packages, i run `rm -rf node_modules && npm install`, and it works fine.

Re (a), i'm not sure what "compatible versions" are, but the JS ecosystem almost exclusively uses semver, and the npm CLI is pretty heavily coupled with semver expectations. When I want to use the latest versions of everything, I run `npx salita -f && npm install`.

Re (b), i'm not sure i understand what you mean about "subpackages" - npm prior to v7 had no concept of workspaces or monorepos.

Re (c), this seems like the same solution as for (a) - `rm -rf node_modules && npm install`.

Re (d), i've never interacted with babel's cache - that seems like something _babel_ should be providing a command for, but the solutions for (a) and (c) seem like they'd cover that if the cache resided inside node_modules.

What's wrong with using `rm -rf`? I'm not sure why `npm install --some-flag` is more elegant than `rm -rf node_modules && npm install` - to me, elegance is "things do one thing well", not "one thing does many things at once" :-)

**Reply by midudev** [reply]
@ljharb maybe you're missing the point... As you could read on the original post I already know about `rm -rf`. If all the comments were to remind me about using `rm -rf`... It's fine that you like it. For me, it's far from ideal specially 

- `npx salita -f && npm install` What's salita? Why should I know about that? I only want to recreate my dependencies by removing node_modules by using my package.json versions that could use caret or major versions. For that I want to use just `npm`, not salita.
- As I mentioned I find `rm -rf` something tied to the specific system and something manual that could go wrong.
- Even if it's the correct answer, finding this to be something developers need to do manually over and over, could be better idea to add to `npm`. Even having to create a package like `salita` for managing upgrading packages must be something to consider.
- "things do one thing well". npm is a node package manager and as such does more than one thing. It's like wondering if `npm install` should have a flag for `--no-save` as you could remove manually these fields on the package.json. Or why npm is auditing my packages? The one thing that should do well is managing my packages. I'm asking if a flag could help this, that's all. I already know it's possible manually.
- @targos added another valid use-case even using lockfiles.

PS: In case you're wondering about Compatible Versioning: https://gitlab.com/staltz/comver It's like Semver but essentially simpler.

**Reply by ljharb** [reply]
I think I _was_ missing the point - it sounds like you simply don't like using `rm -rf node_modules`, you don't have a lockfile (with a lockfile, `npm ci` wipes node_modules on install, and `rm package-lock.json && npm install` would generate a fresh fully up to date lockfile), and you want a flag on `npm install` that does it for you?

**Comment by MylesBorins** [comment, marked as answer]
Hey @midudev thanks for opening the discussion. As mentioned in the discussion already the behavior you are asking for is something that is already done via `npm ci`, but I see you mention that your company doesn't use lock files.

If you would like to propose a new CLI feature the best next step would be to open an issue on https://github.com/npm/rfcs clearly outlining your proposal so you can get direct feedback from the CLI team.
